### PR TITLE
Expose a conduit list s3 operation

### LINKED ISSUE #173: Expose conduit for 'list'

### PR DESCRIPTION
**Pull request by nhibberd**
Fixes #173

### CODE REVIEW

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 1]
```diff
@@ -280,10 +281,23 @@ listObjects :: Address -> AWS ([Address], [Address])
...
 
--- list the addresses, keys first, then prefixes
 list :: Address -> AWS [Address]
 list a =
-  (\(p, k) -> k <> p) <$> listObjects a
+  list' a >>= ($$ DC.consume)
+
+list' :: Address -> AWS (Source AWS Address)
+list' a@(Address (Bucket b) (Key k)) = do
+  let pp kk = if T.null kk then "" else if T.isSuffixOf "/" kk then kk else kk <> "/"
```
Comment (markhibberd) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Is this different to `<>` if so, should `<>` change? Either way probably good to have a test of some-sort to sanity check this.

**Comment by nhibberd** [comment]
@EduardSergeev bump

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 2]
```diff
@@ -269,7 +283,7 @@ downloadWithMode mode a f = do
   unlessM (exists a) . fail $ "Can not download when the source does not exist [" <> (T.unpack $ addressToText a) <> "]."
   liftIO $ createDirectoryIfMissing True (dropFileName f)
   r <- getObject' a
-  r' <- maybe (fail "shit") pure r
+  r' <- maybe (fail $ "Unable to retrieve content for source [" <> (T.unpack $ addressToText a) <> "].") pure r
   liftIO . runResourceT . ($$+- sinkFile f) $ r' ^. gorsBody ^. to bodyResponse
```
Comment (EduardSergeev) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Maybe it's better to use exception here istead of 'fail'? To be in line with other functions in mismi.

Comment (nhibberd) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Yeah looks like `downloadWIthMode` might have been missed in the cleanup, still has a number of fails

**Comment by nhibberd** [comment]
@EduardSergeev Updated error types

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 3]
```diff
@@ -255,21 +256,34 @@ listObjects :: Address -> AWS ([Address], [Address])
...
 downloadWithMode :: WriteMode -> Address -> FilePath -> AWS ()
 downloadWithMode mode a f = do
-  when (mode == Fail) . whenM (liftIO $ doesFileExist f) . fail $ "Can not download to a target that already exists [" <> f <> "]."
-  unlessM (exists a) . fail $ "Can not download when the source does not exist [" <> (T.unpack $ addressToText a) <> "]."
+  when (mode == Fail) . whenM (liftIO $ doesFileExist f) . throwM $ DestinationFileExists f
+  unlessM (exists a) . throwM $ SourceMissing DownloadError a
   liftIO $ createDirectoryIfMissing True (dropFileName f)
   r <- getObject' a
-  r' <- maybe (fail "shit") pure r
+  r' <- maybe (throwM $ NoContent a) pure r
```
Comment (EduardSergeev) [inline mismi-s3/src/Mismi/S3/Commands.hs]: What if we just call `getObject'` omitting `exists` and depending on the result we can either `throwM SourceMissing` or return the result with `to bodyResponse`. This way we don't need `NoContent` exception?

**Comment by EduardSergeev** [comment]
:+1:

**Comment by nhibberd** [comment]
Fixes #173

**Pull request merged**